src/Accounts/Authenticators/MsalAccessToken.cs

```diff
@@ -39,50 +39,79 @@ public class MsalAccessToken : IAccessToken
         {
+            TokenCredential = tokenCredential;
+            TokenRequestContext = tokenRequestContext;
             AccessToken = token;
+            ExpiredOn = expiresOn;
             UserId = userId;
             TenantId = tenantId;
             HomeAccountId = homeAccountId;
         }
 
         public void AuthorizeRequest(Action<string, string> authTokenSetter)
         {
+            Renew();
             authTokenSetter("Bearer", AccessToken);
         }
 
         public static async Task<IAccessToken> GetAccessTokenAsync(
-            ValueTask<AccessToken> result,
+            TokenCredential tokenCredential,
+            TokenRequestContext requestContext,
+            CancellationToken cancellationToken,
             string tenantId = null,
             string userId = null,
             string homeAccountId = "")
         {
-            var token = await result;
-            return new MsalAccessToken(token.Token, tenantId, userId, homeAccountId);
+            var token = await tokenCredential.GetTokenAsync(requestContext, cancellationToken);
+            return new MsalAccessToken(tokenCredential, requestContext, token.Token, token.ExpiresOn, tenantId, userId, homeAccountId);
         }
 
-        public static async Task<IAccessToken> GetAccessTokenAsync(
-            ValueTask<AccessToken> result,
-            Action action,
-            string tenantId = null,
-            string userId = null)
-        {
-            var token = await result;
-            action();
-            return new MsalAccessToken(token.Token, tenantId, userId);
-        }
 
         public static async Task<IAccessToken> GetAccessTokenAsync(
             Task<AuthenticationRecord> authTask,
-            Func<ValueTask<AccessToken>> getTokenAction,
-            CancellationToken cancellationToken = default(CancellationToken))
+            TokenCredential tokenCredential,
+            TokenRequestContext requestContext,
+            CancellationToken cancellationToken)
         {
             var record = await authTask;
             cancellationToken.ThrowIfCancellationRequested();
-            var token = await getTokenAction();
+            var token = await tokenCredential.GetTokenAsync(requestContext, cancellationToken);
 
-            return new MsalAccessToken(token.Token, record.TenantId, record.Username, record.HomeAccountId);
+            return new MsalAccessToken(tokenCredential, requestContext, token.Token, token.ExpiresOn, record.TenantId, record.Username, record.HomeAccountId);
+        }
+
+
+        private void Renew()
+        {
+            if(IsExpired())
+            {
+                var token = TokenCredential.GetToken(TokenRequestContext, default(CancellationToken));
+                AccessToken = token.Token;
+                ExpiredOn = token.ExpiresOn;
+            }
+        }
+
+        private bool IsExpired()
```

Comment:
Suggest rename to match the inner logic
```suggestion
        private bool IsNearExpiration()
```